File: 2024/day8.rb
def get_antinode_pos(a1, a2, n)
  [a1[0] + n * (a1[0] - a2[0]), a1[1] + n * (a1[1] - a2[1])]
end

def in_bounds(pos, size)
  pos[0] >= 0 && pos[0] < size && pos[1] >= 0 && pos[1] < size
end

def get_antinodes_in_bounds(a1, a2, size, maxn)
  n = 1
  antinodes = Set.new
  while n <= maxn do
    antinode = get_antinode_pos(a1, a2, n)
    break unless in_bounds(antinode, size)
    antinodes << antinode
    n += 1
  end
  antinodes
end

def count_antinodes(grid, part2)
  size = grid.size
  freqs = Hash.new { |h, k| h[k] = [] }
  grid.each_with_index do |row, y|
    row.each_with_index do |c, x|
      freqs[c] << [y, x] if c != '.'
    end
  end

  antinodes = Set.new
  freqs.each do |f, antenna|
    antenna.combination(2).each do |a1, a2|
      maxn = part2 ? 1000 : 1
      antinodes += [a1, a2] if part2
      antinodes += get_antinodes_in_bounds(a1, a2, size, maxn)
      antinodes += get_antinodes_in_bounds(a2, a1, size, maxn)
    end
  end

  antinodes.size
end

grid = STDIN.read.split("\n").map(&:chars)
puts count_antinodes(grid, false)
puts count_antinodes(grid, true)

File: 2024/day8.hs
import Data.List qualified as List
import Data.Map qualified as Map
import Data.Set qualified as Set

type Pos = (Int, Int)

getAntinodePos :: Int -> Pos -> Pos -> Pos
getAntinodePos n (a1y, a1x) (a2y, a2x) = (a1y + n * (a1y - a2y), a1x + n * (a1x - a2x))

inBounds :: Int -> Pos -> Bool
inBounds size (y, x) = y >= 0 && y < size && x >= 0 && x < size

getAntinodesInBounds :: Int -> Int -> Pos -> Pos -> Set.Set Pos
getAntinodesInBounds size maxn a1 a2 = do
  Set.fromList $ antinodes a1 a2 ++ antinodes a2 a1
  where
    antinodes b1 b2 =
      takeWhile (inBounds size) [getAntinodePos n b1 b2 | n <- [1 .. maxn]]

combinations :: Int -> [a] -> [[a]]
combinations k ns = filter ((k ==) . length) $ List.subsequences ns

enumerateGrid :: [[Char]] -> [(Char, Pos)]
enumerateGrid grid = do
  foldl1 (++) $
    zipWith
      (\r yidx -> zipWith (\a xidx -> (a, (yidx, xidx))) r ([0 ..] :: [Int]))
      grid
      ([0 ..] :: [Int])

countAntinodes :: [[Char]] -> Bool -> Int
countAntinodes grid part2 = do
  let maxn = if part2 then 100 else 1
  let freqs = Map.fromListWith (++) $ map (\(k, a) -> (k, [a])) $ filter (\(c, _) -> c /= '.') $ enumerateGrid grid
  let nodePairs = foldl1 (++) $ map (combinations 2 . snd) $ Map.toList freqs
  let antinodes = foldl1 Set.union $ map (\[a1, a2] -> getAntinodesInBounds (length grid) maxn a1 a2) nodePairs
  let allAntinodes = if part2 then Set.union antinodes $ Set.fromList $ foldl1 (++) nodePairs else antinodes
  length allAntinodes

main = do
  contents <- getContents
  print $ countAntinodes (lines contents) False
  print $ countAntinodes (lines contents) True
